Update Capacity Test to optionally test online learning

rmccall84 opened the pull request:
Fixes https://github.com/numenta/nupic.research/issues/134

rmccall84 commented:
@chetan51 @scottpurdy FYI, I've been updating this PR as I make additions, refactors, renames to the experiment-related code. It doesn't have to be reviewed atm.

scottpurdy commented:
@rmccall84 - I'd highly recommend that you get stuff merged in iteratively.If you can break this up into multiple PRs to get started that would be great. Regardless, better to get it merged now that get a backlog of changes.

rmccall84 commented:
@chetan51 I'd like you to please review this. We can go over it together if you'd like.

chetan51 commented on sensorimotor/experiments/capacity/data_utils.py:
```diff
@@ -0,0 +1,186 @@
+  Plots the data produced by
+  sensorimotor/experiments/capacity/run.py
+  """
+  filesDir = "/Users/rmccall/nta/nupic.research/sensorimotor/experiments" \
```
This should be an argument to this script.

chetan51 commented on sensorimotor/experiments/capacity/data_utils.py:
```diff
@@ -0,0 +1,186 @@
+  # metrics:
+  # Mean & Max Stability, Mean & Max Distinctness, Mean & Max Bursting Cols
+  yColumnIdxs = [11, 9, 16, 14, 46, 44]
+  yStdDevIdxs = [12, -1, 17, -1, 47, -1]
```
What does `-1` mean here?

chetan51 commented on sensorimotor/experiments/capacity/data_utils.py:
```diff
@@ -0,0 +1,186 @@
+  """
+  filesDir = "/Users/rmccall/nta/nupic.research/sensorimotor/experiments" \
+             "/capacity/output/varyWorlds-reasonable/slow3x"
+  combinedFileName = "allCombined.csv"
```
This too.

chetan51 commented on sensorimotor/experiments/capacity/run.py:
```diff
@@ -69,12 +69,14 @@ def run(args):
   numWorlds = defn["worlds"]
   numElements = defn["elements"]
 
-  filePrefix = os.path.join(outputDir, "{0}x{1}".format(numWorlds,
```
If you instead use the format flag to pre-fill the two numbers with `0`s, you can avoid the complicated logic below.

chetan51 commented on sensorimotor/sensorimotor/sensorimotor_experiment_runner.py:
```diff
@@ -149,8 +151,9 @@ def feedLayers(self, sequences, tmLearn=True, tpLearn=None, verbosity=0,
 
-    self.tm.mmClearHistory()
-    self.tp.mmClearHistory()
+    if clearHistory:
```
Remove `clearHistory` param and move following two lines to `runTest` in `experiment.py`.

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -47,168 +47,233 @@
 
-  # Initialize experiment
-  start = time.time()
+# Default number of true input bits
```
`true` => `on`

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -47,168 +47,233 @@
-
-  csvFilePath = os.path.join(outputDir, "{0}x{1}.csv".format(numWorlds,
-                                                             numElements))
 
```
Remove extra line

rmccall84 replied:
What line? your comment is in a deleted section?

chetan51 replied:
My  comment is on an empty line actually

chetan51 replied:
This one: https://github.com/numenta/nupic.research/pull/135/files#diff-3471a27934a13cf92b8c665a86d6e504R60

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -47,168 +47,233 @@
-  start = time.time()
+# Default number of true input bits
+W = 20
+DEFAULTS = {"n": N,
```
Remove and have this script accept a params file

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -47,168 +47,233 @@
+                         "spVerbosity": 0}}
+
+IS_ONLINE_LEARNING = True
+PLOT = 2
```
Accept as optional arguments to this script that override these defaults (and reset these defaults to their original values)

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -47,168 +47,233 @@
+# most recent
+TRAINING_CLEAR_HISTORY = False
+SHOW_PROGRESS_INTERVAL = 300
+TWOPASS_TM_TRAINING_SWEEPS = 2
```
Rename `sweeps` to `reps` everywhere

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -47,168 +47,233 @@
+            completeSequenceLength):
+  print "Testing (worlds: {0}, elements: {1})...".format(numWorlds,
+                                                         numElements)
+  numSequences = 4
```
Move to constant

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -47,168 +47,233 @@
+  shading = 0.4
+  if plotVerbosity >= 1:
+    # title = "worlds: {0}, elements: {1}".format(numWorlds, numElems)
+    title = "Online: " + str(IS_ONLINE_LEARNING) + ", Training phase"
```
Combine above two lines

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -47,168 +47,233 @@
+  if plotVerbosity >= 1:
+    # title = "worlds: {0}, elements: {1}".format(numWorlds, numElems)
+    title = "Online: " + str(IS_ONLINE_LEARNING) + ", Training phase"
+    # runner.tp.mmGetPlotConnectionsPerColumn(title=title)
```
Uncomment

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -47,168 +47,233 @@
-                 ["min", "max", "sum", "mean", "stddev"]]
-      row += [metric.min, metric.max, metric.sum,
-              metric.mean,  metric.standardDeviation]
+  if "isOnline" in params:
```
Make this required

chetan51 commented on sensorimotor/experiments/capacity/params/reasonable_slow/1024_10x.yaml:
```diff
@@ -0,0 +1,24 @@
+n: 1024
```
Rename these param files

chetan51 commented on sensorimotor/experiments/capacity/data_utils.py:
```diff
@@ -0,0 +1,186 @@
+
+
+
+# TODO: Discuss the possibility of an abstracted (and corrected) form of this
```
Probably not necessary

rmccall84 commented:
@chetan51 when you have the time, would you please look at this again?

chetan51 commented on sensorimotor/experiments/capacity/defs/varyElements.yaml:
```diff
@@ -0,0 +1,16 @@
+- worlds: 2
```
File names should be underscore_case

chetan51 commented on sensorimotor/experiments/capacity/defs/varyElements.yaml:
```diff
@@ -0,0 +1,16 @@
+- worlds: 2
+  elements: 80
```
Do we want these defs to go up to 100?

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -44,179 +44,252 @@
+  print "Training temporal pooler..."
+
+  runner.tm.mmClearHistory()
+  runner.tp.mmClearHistory()
```
Minor: add newline after this

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -44,179 +44,252 @@
+                 completeSequenceLength, verbosity):
+  print "Testing (worlds: {0}, elements: {1})...".format(numWorlds, numElements)
+  runner.tm.mmClearHistory()
+  runner.tp.mmClearHistory()
```
Minor: add newline after this

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -44,179 +44,252 @@
+
+def runTestPhase(runner, randomAgents, numWorlds, numElements,
+                 completeSequenceLength, verbosity):
+  print "Testing (worlds: {0}, elements: {1})...".format(numWorlds, numElements)
```
Minor: add newline after this

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -44,179 +44,252 @@
+
+def run(numWorlds, numElems, paramsPath, outputDir, plotVerbosity,
+        consoleVerbosity, params=None):
+  if params is None:
```
Do this in `__main__` instead, remove `paramsPath`, and require `params` to be passed in to this function.

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -44,179 +44,252 @@
+  tmParams = params["tmParams"]
+  tpParams = params["tpParams"]
+  isOnline = params["isOnline"]
+  onlineTrainingReps = params["onlineTrainingReps"] if isOnline else "N/A"
```
If possible, use `None` instead of `"N/A"`.

chetan51 commented on sensorimotor/experiments/capacity/experiment.py:
```diff
@@ -44,179 +44,252 @@
 
 
 
+parser = OptionParser(usage="%prog NUM_WORLDS NUM_ELEMENTS PARAMS_DIR "
```
Isn't this usually under `__main__`?

chetan51 commented on sensorimotor/experiments/capacity/params/reasonable/2048.yaml:
```diff
@@ -1,5 +1,7 @@
 n: 2048
```
What happened to the offline params?

rmccall84 replied:
Offline params? There is a boolean 'isOnline' param in the yamls.

chetan51 replied:
It seems as if all the def files are now online learning.

rmccall84 replied:
Do you want two files, one for online=True and another for online=False? I suppose we might not even want online=False in the near future, but for now I'd like to add this.

chetan51 replied:
Yes, I thought we discussed having two sets of directories, one top level directory called `online` and one called `offline`, with param files in those directory structures. Then when we want to get rid of `offline`, we can just delete that directory (and remove the `isOnline` parameter).

Otherwise, how can you run the offline parameters to test currently?

chetan51 commented on sensorimotor/experiments/capacity/run.py:
```diff
@@ -32,55 +32,55 @@
-  filePrefix = os.path.join(outputDir, "{0}x{1}".format(numWorlds,
-                                                        numElements))
+  print "Starting {0} worlds x {1} elems...".format(numWorlds, numElements)
+  fileName = "{0}x{1}".format(numWorlds, numElements)
```
Doesn't this break the alphabetical sorting of files?

rmccall84 commented:
@chetan51 with the plot PR merged we can move ahead here. I've addressed your comments in the recent commits "Fixing defs and params argument handling in experiment.py" and "Online and offline parameter directories." Many thanks

chetan51 commented on sensorimotor/experiments/capacity/defs/vary_elements.yaml:
```diff
@@ -0,0 +1,20 @@
+- worlds: 2
```
Can you reverse the order of these experiments, so that the fastest, smallest ones run first? Same for `vary_worlds.yml`.

rmccall84 replied:
I found that experiments run faster when you run the larger conditions first. The largest condition ends up taking longer than several of the smaller conditions.  If you'd still prefer smaller first I'll change it.

chetan51 commented:
Other than the minor suggestion above, looks great! Feel free to merge after that change, thanks.

rmccall84 commented:
@chetan51 I don't have write access. So would you plz merge it?
